### DISCUSSION TITLE
List connected clients

### DISCUSSION THREAD

**Discussion by Charlyx12** [FreeOpcUa/opcua-asyncio#845] [Q&A] [answered]
Hi, is there a way to list all the clients that are connected to the server? I mean, their names for example.

**Comment by AndreasHeine** [comment, marked as answer]
probably the binary server:
https://github.com/FreeOpcUa/opcua-asyncio/blob/729ef892f084679a955f8be423ab5b3400a0902a/asyncua/server/server.py#L90

https://github.com/FreeOpcUa/opcua-asyncio/blob/729ef892f084679a955f8be423ab5b3400a0902a/asyncua/server/binary_server_asyncio.py#L112

**Reply by Charlyx12** [reply]
Thanks for the answer! I tried that but it shows tons of information and im just interested in taking the names :(

**Reply by Charlyx12** [reply]
For example, in this case theres a client connected to the server, when I print doing `print(server.bserver.clients)` i get the following result:
```
[OPCUAProtocol(('127.0.0.1', 56703), InternalSession(name:('127.0.0.1', 56703), user:User(role=<UserRole.Anonymous: 1>, name=None), id:NodeId(Identifier=11, NamespaceIndex=0, NodeIdType=<NodeIdType.TwoByte: 0>), auth_token:NodeId(Identifier=1001, NamespaceIndex=0, NodeIdType=<NodeIdType.FourByte: 1>)))]
```
I had the idea to obtain the names by processing the result as a string and searching for the 'name' field, but i dont think its a great solution. Is there another way to do that or do you think this could be a possible solution?

Thanks

**Reply by swamper123** [reply]
Either you stringify the result and use regex or you have to create a function, which filters out the informations you want.

**Reply by Charlyx12** [reply]
ok, thank you so much. :)

**Reply by swamper123** [reply]
Either way, I would recommened that you should use 3 clients, if you want to built such a function. This avoids pitfalls.

**Reply by Charlyx12** [reply]
I was able to get the names by using regex:
`name=re.findall("(?<=name\=\')(\w+)",str(server.bserver.clients[server.bserver.clients.index(cli)]))`
where cli is the iterator variable used in a for loop to move through _server.bserver.clients_ list

However, I found another way to get the same results:
`name=server.bserver.clients[server.bserver.clients.index(cli)].processor.session.user.name`

If you are also interested in getting the role, you could use:
`role = server.bserver.clients[server.bserver.clients.index(cli)].processor.session.user.role.name`
